### PR TITLE
Add cargo-risczero install

### LINKED ISSUE #754: Guest builds fail due to `cc` crate 1.0.80
After the release of `cc` crate 1.0.80 the zkvm guest build fails with the following errors:

```
method_name: error[E0152]: duplicate lang item in crate `core` (which `rustc_std_workspace_core` depends on): `sized`.
method_name:   |
method_name:   = note: the lang item is first defined in crate `core` (which `std` depends on)
method_name:   = note: first definition in `core` loaded from /home/user/.rustup/toolchains/nightly-2023-03-06-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-05898138a596088a.rlib
method_name:   = note: second definition in `core` loaded from /home/user/tmp/cc-bug/target/riscv-guest/release/deps/libcore-d6bc9e683b556e25.rmeta
<snip>
```

## Workaround:

Place the following config:
```toml
[patch.crates-io]
cc = { git = "https://github.com/rust-lang/cc-rs", rev = "e5bbdfa" }
```
into either:
* Your current rust workspace root `.cargo/config.toml`
* Or system wide at `~/.cargo/config.toml`

## Origin

Appears to be related to the following commit https://github.com/rust-lang/cc-rs/commit/17c8858efbeec6094c21973e3d2e9773916b6aa2 and respective PR in `cc-rs`: https://github.com/rust-lang/cc-rs/pull/780

`cc` is not in the default guest cargo depenendency tree but is in the `build-std` build tree so its most likely a mix up within our `std` building.

### PR DESCRIPTION
**Pull request by flaub**
Use the `cargo-risczero` tool to install the `risv32im-risc0-zkvm-elf` toolchain to avoid the use of the unstable `-Z build-std` cargo feature. This fixes #754.

This change means that a rustup toolchain named `risc0` must be installed for `risc0-build` to work. To make installing this toolchain easier, the `cargo-risczero` tool has been extended to have the following commands:

* `install`: This will download the latest rust toolchain published as a release in https://github.com/risc0/rust/releases.  It will then use `rustup toolchain link risc0 $SYSROOT` to make the toolchain available to rustup. `risc0-build` will run `cargo +risc0 build ...` in order to select this new toolchain.
* `build-toolchain`: This command will download the latest source of the rust fork and build it locally. It will then install it using `rustup toolchain link`. This command is useful for building and publishing a new release of the fork. It can also be used to build the toolchain on a host that we don't have a prebuild archive published.

### CODE REVIEW

File: risc0/build/src/lib.rs [hunk 1]
```diff
@@ -362,21 +202,13 @@ fn build_guest_package<P>(
...
+        "+risc0",
         "build",
         "--release",
         "--target",
-        guest_build_env.target_spec.to_str().unwrap(),
-        "-Z",
-        build_std.as_str(),
-        "-Z",
-        "build-std-features=compiler-builtins-mem",
+        "riscv32im-risc0-zkvm-elf",
```
Comment (mothran) [inline risc0/build/src/lib.rs:211]: By dropping these -Z's what are the remaining bits keeping us on nightly?

File: risc0/cargo-risczero/src/commands/install.rs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+/// Release asset returned by Github API.
+#[derive(Deserialize)]
+struct GithubAsset {
+    browser_download_url: String,
+    name: String,
+}
+
+impl Install {
+    pub fn run(&self) -> Result<()> {
```
Comment (mothran) [inline risc0/cargo-risczero/src/commands/install.rs:50]: Should we document or provide a method for `uninstall` ?

**Review by mothran — APPROVED** [review]
I left a few general questions / possible improvements but I think this is a solid foundation

**Review by SchmErik — APPROVED** [review]

File: risc0/cargo-risczero/src/commands/install.rs [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    if let Some(token) = gh_token {
+        headers.insert("authorization", format!("Bearer {token}").parse()?);
+    }
+
+    let client = Client::builder()
+        .default_headers(headers)
+        .user_agent("cargo-wasix")
+        .build()?;
+
+    let repo = RUST_REPO
```
[the remark this answers is not in the document]
Comment (flaub) [reply risc0/cargo-risczero/src/commands/install.rs:126]: Soon, yes :)

It's based on: https://github.com/wasix-org/rust/tree/wasix/.github/workflows

File: risc0/build/src/lib.rs (hunk 1 above)
Comment (flaub) [reply risc0/build/src/lib.rs:213]: We need to get a build of `cargo` which is independent of the host toolchain.

File: risc0/cargo-risczero/src/commands/build_toolchain.rs [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        RustupToolchain::link(RUSTUP_TOOLCHAIN_NAME, &out.toolchain_dir)?;
+
+        Ok(())
+    }
+
+    /// Initialize a Git repo.
+    ///
+    /// Clone if it doesn't exist yet, otherwise update the branch/tag.
+    fn prepare_git_repo(&self, source: &str, tag: &str, path: &Path) -> Result<()> {
+        eprintln!("Preparing git repo {source} with tag/branch {tag}");
```
[the remark this answers is not in the document]
Comment (flaub) [reply risc0/cargo-risczero/src/commands/build_toolchain.rs:67]: I think this is less about tracing and more about interacting with the user. I could see having both (one for diagnostics, one for end users).

[the remark this answers is not in the document]
Comment (mothran) [reply risc0/cargo-risczero/src/commands/build_toolchain.rs:67]: Good point, does make sense to force eprint the things directly that need to be seen.

File: risc0/cargo-risczero/src/commands/install.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mothran) [reply risc0/cargo-risczero/src/commands/install.rs:126]: Great!

**Review by tzerrell — APPROVED** [review]
:tada:

**Pull request merged**